composer/loggers/in_memory_logger.py

```diff
@@ -74,3 +75,50 @@ def log_metric(self, timestamp: Timestamp, log_level: LogLevel, data: TLogData):
             self.data[k].append((timestamp, log_level, v))
         self.most_recent_values.update(data)
         self.most_recent_timestamps.update({k: timestamp for k in data})
+
+    def get_timeseries(self, metric: str, time_unit: Union[str, None] = None):
+        """Returns logged data as dict containing values of a desired metric over time.
+
+        Args:
+            metric (str): Metric of interest. Must be present in self.data.keys().
+            time_unit(str or None, optional) Which unit of time to return. Options =
+                {"batch", "epoch", "sample", "token", "duration", None}. If ``None``,
+                will return all available units of time. Default=``None``.
+
+        Returns:
+            timeseries (dict): Dictionary in which one key is ``metric``, and the
+                associated value is a list of values of that metric. The remaining key(s)
+                are each a unit of time, and the associated values are each a list of
+                values of that time unit for the corresponding index of the metric. For
+                example, plotting timeseries['accuracy/val'] vs. timeseries['epoch'] would
+                plot validation accuracy as a function of epoch. And if
+                timeseries['accuracy/val'][5] == 41.2, and timeseries['batch'][5] == 680,
+                that means that 'accuracy/val' was equal to 41.2 on the 680th batch.
+        """
+
+        # Check time_unit string
+        valid_time_strings = {"batch", "epoch", "sample", "token", "duration", None}
+        if time_unit not in valid_time_strings:
+            raise ValueError(f"InMemoryLogger.get_timeseries() passed {time_unit} "
+                             f"as argument for parameter `time_unit`. This method only accepts "
+                             f"the following arguments: {valid_time_strings}.")
+
+        # Check that desired metric is in present data
+        if metric not in self.data.keys():
+            raise ValueError(f"Invalid value for argument `metric`: {metric}. Requested "
+                             "metric is not present in self.data.keys().")
+
+        timeseries = defaultdict(list)
+        # Iterate through datapoints
+        for datapoint in self.data[metric]:
+            metric_value = datapoint[2]
+            timeseries[metric].append(metric_value)
+            timestamp = datapoint[0]
+            if time_unit:
+                timeseries[time_unit].append(getattr(timestamp, time_unit).value)
+            else:
+                # Iterate through time units and add them all!
+                for field in timestamp._fields:
+                    time_value = getattr(timestamp, field).value
+                    timeseries[field].append(time_value)
+        return timeseries
```

Comment:
Having type check problems here: `Expression of type "defaultdict[Unknown, list[Unknown]]" cannot be assigned to return type "Dict[str, TLogData]"`. Suggested solutions @ravi-mosaicml, @hanlint?

Reply:
Easiest way to fix would be to not use a defaultdict and instead use a dict